legend-sdlc-server-fs/src/main/java/org/finos/legend/sdlc/server/api/workspace/FileSystemWorkspaceApi.java

```diff
@@ -0,0 +1,238 @@
+            for (Ref branch : allBranches)
+            {
+                String type = repository.getConfig().getString("branch", Repository.shortenRefName(branch.getName()), "type");
+                if (type != null && type.equals(branchType))
+                {
+                    branchesOfType.add(branch);
+                }
+            }
+        }
+        catch (GitAPIException e)
+        {
+            e.printStackTrace();
+        }
+        return branchesOfType.stream().map(branch -> workspaceBranchToWorkspace(projectId, null, branch, wType, null)).collect(Collectors.toList());
+    }
+
+    private static Workspace workspaceBranchToWorkspace(String projectId, VersionId patchReleaseVersionId, Ref branch, WorkspaceType workspaceType, ProjectFileAccessProvider.WorkspaceAccessType workspaceAccessType)
+    {
+        return (branch == null) ? null : fromWorkspaceBranchName(projectId, patchReleaseVersionId, branch, workspaceType, workspaceAccessType);
+    }
+
+    protected static Workspace fromWorkspaceBranchName(String projectId, VersionId patchReleaseVersionId, Ref branch, WorkspaceType workspaceType, ProjectFileAccessProvider.WorkspaceAccessType workspaceAccessType)
+    {
+        String userId = workspaceType == WorkspaceType.GROUP ? null : "user";
+        return new Workspace()
+        {
+            @Override
+            public String getProjectId()
+            {
+                return projectId;
+            }
+
+            @Override
+            public String getUserId()
+            {
+                return userId;
+            }
+
+            @Override
+            public String getWorkspaceId()
+            {
+                return Repository.shortenRefName(branch.getName());
+            }
+        };
+    }
+
+    public static  Repository retrieveRepo(String projectId) throws IOException
+    {
+        File[] repoDirs = new File(FSConfiguration.getRootDirectory()).listFiles(File::isDirectory);
+        if (repoDirs != null)
+        {
+            for (File repoDir : repoDirs)
+            {
+                Repository repository = FileRepositoryBuilder.create(new File(repoDir, ".git"));
+                String storedProjectId = repository.getConfig().getString("project", null, "id");
+                if (projectId.equals(storedProjectId))
+                {
+                    return repository;
+                }
+            }
+        }
+        return null;
+    }
+
+    public static Ref getGitBranch(String projectId, String workspaceId)
+    {
+        try
+        {
+            Repository repository = retrieveRepo(projectId);
+            List<Ref> allBranches = Git.wrap(repository).branchList().setListMode(ListBranchCommand.ListMode.ALL).call();
+            for (Ref branch : allBranches)
+            {
+                if (Repository.shortenRefName(branch.getName()).equals(workspaceId))
+                {
+                    return branch;
+                }
+            }
```

Comment:
From a workspace specification, you should have enough information to construct an exact branch name. You should not have to search through all branch names for a repo.